### ISSUE TITLE
amigaos4 compile

### ISSUE THREAD

**Issue by kas1e** [sodero/MUI-Vim#5]
Howdy !

Tried to build amigaos4 version from root's makefile on cross-compiler as you say before, with using of "make UNM=AmigaOS4 ARC=powerpc". And that one fail, it didn't use ppc-amigaos-gcc, but just pure gcc:

~~~
$ cd /amiga/vimmui_test/
$ make UNM=AmigaOS4 ARC=powerpc
make -C src -f Make_ami.mak PATCHLEVEL=`cat src/.pat`
make[1]: Entering directory '/amiga/vimmui_test/src'
gcc -c -DNO_ARP -DUSE_TMPNAM -DHAVE_STDARG_H -DHAVE_TGETENT -DHAVE_TERMCAP -DFEAT_GUI -DFEAT_GUI_MUI -DFEAT_HUGE -DFEAT_BROWSE -DFEAT_TOOLBAR -I proto -Wno-attributes -Wextra -O3 -DPATCHLEVEL=\"1317\"   -c -o arabic.o arabic.c
~~~

I know why it happens :  you do change Installer/Makefile. Not the root's makefile, and not the src/Make_ami.mak.

Are you sure we still need all that "installer" problems ?  I know you say you love to tinker with them, but its all just make whole thing be more harder and unnecessary-steps-for-users. If you doesnt mind, for os4 build it will be for sure much better to avoid installer at all (it looks ugly on os4, as your scripts use some original installer, while os4 have new , python based installer utility, which looks fancy when some big heavy software need it, but code-wise will be totally different than yours of course).

I mean, i can just commit my makefile to "src" directory, and anyone can just build it by pure going to src directory and type "make -f Make_os4_clib2.mak".

Tell me what you think plz about

**Comment by sodero** [comment]
Hi Kas1e!

2019-05-27 9:54 GMT+02:00, kas1e <notifications@github.com>:
> Howdy !
>
> Tried to build amigaos4 version from root's makefile on cross-compiler as
> you say before, with using of "make UNM=AmigaOS4 ARC=powerpc". And that one
> fail, it didn't use ppc-amigaos-gcc, but just pure gcc:
>
> ~~~
> $ cd /amiga/vimmui_test/
> $ make UNM=AmigaOS4 ARC=powerpc
> make -C src -f Make_ami.mak PATCHLEVEL=`cat src/.pat`
> make[1]: Entering directory '/amiga/vimmui_test/src'
> gcc -c -DNO_ARP -DUSE_TMPNAM -DHAVE_STDARG_H -DHAVE_TGETENT -DHAVE_TERMCAP
> -DFEAT_GUI -DFEAT_GUI_MUI -DFEAT_HUGE -DFEAT_BROWSE -DFEAT_TOOLBAR -I proto
> -Wno-attributes -Wextra -O3 -DPATCHLEVEL=\"1317\"   -c -o arabic.o arabic.c
> ~~~
>
> I know why it happens :  you do change Installer/Makefile. Not the root's
> makefile, and not the src/Make_ami.mak.

Yes, but only the 'installer' target uses this, it's used to generate
the aminet readme and so on. The building of Vim itself work without
these variables.

>
> Are you sure we still need all that "installer" problems ?  I know you say
> you love to tinker with them, but its all just make whole thing be more
> harder and unnecessary-steps-for-users. If you doesnt mind, for os4 build it
> will be for sure much better to avoid installer at all (it looks ugly on
> os4, as your scripts use some original installer, while os4 have new ,
> python based installer utility, which looks fancy when some big heavy
> software need it, but code-wise will be totally different than yours of
> course).

Ok, I won't clinch to the installer idea anymore, let's get rid of it,

>
> I mean, i can just commit my makefile to "src" directory, and anyone can
> just build it by pure going to src directory and type "make -f
> Make_os4_clib2.mak".

Sure, that will work. I think it's nice to have support for aminet
packaging (no installer ;)) too, but I can fix that.

Regards,
Ola

**Comment by kas1e** [comment]
Do you mean we will use my makefile for os4, or you think we can fix root's makefile ?

**Comment by sodero** [comment]
We can use your makefile.

**Comment by kas1e** [comment]
Ok i added most of amigaos4 changes, as well as makefile, as well as some general fixes. While i am almost sure everything will be fine for morphos/aros in terms of compilation after my changes, plz, check if all is ok still. Expectually i not 100% sure about my last commit about proper filling of EasyStruct structure (to avoid warning and possible bugs). 

Plz check if everything still compiles fine, so i can continue with commits futher.

Thanks !

**Comment by sodero** [comment]
Excellent! I needed to ifdef some of the EasyStruct things but now everything builds and runs on AROS (and then it should work on MorphOS too).

**Comment by kas1e** [comment]
Cool ! With latest commit i add new objects to aos4 makefile, so it now fully builds. 

Through, i do not know at moment what happens, but while all runs fine, shows fine, any opening of .txt file cause crash of vim in the copy_viminfo_marks().

Is it works for you on morphos/aros ?

**Comment by sodero** [comment]
Yes, it works on AROS. 

Are you running it from the shell or from WB? 

I've had problems when running it without the script from WB, so I think I have to put that one back, unfortunately. The input / output handling in os_amiga.c is very confusing to me so I don't dare (or have the energy) to touch it right now.

**Comment by kas1e** [comment]
Its and from shell and from wb, and with gvim script and without.. I use data files from 8.1 version, but dunno if it can be related..

**Comment by sodero** [comment]
I think the data files should be OK. But you could try to delete VIM:viminfo (given the name of the function above) and see what happens.

**Comment by sodero** [comment]
Or even remove the VIM: assign and start it completely blank...

**Comment by kas1e** [comment]
Yeah, deleting viminfo from S: help. Maybe it was corrupted or something ..

Anyway, now all runs again, will continue to add other fixes.

Also have question: where is better place for us to place stack cookie in the vim ? Usually it sets right before main() function called.

I want to set 1mb of default stack size, so no plugins/etc will ever crashes, and while on aos4 we have that "stack cookie" which can overwrite stack size of icon, stack size setting in system/shell, etc, dunno if morphos and aros have the same ? If i remember right morphos have (at least in odyssey and vpdf source code i found setting of stack size for morphos).

**Comment by sodero** [comment]
Nice that it works now. I don't think it matters where we put it, it's probably treated like the version string. My suggestion is to put it in os_amiga.c before the version string. 1M is a good size, should be enough, and it will still be insignificant on an NG system.

**Comment by sodero** [comment]
I'll try to find out how stack cookies work on AROS / MOS.

**Comment by kas1e** [comment]
Ok, will check now if it will works from os_amiga.c, and then will commit it

**Comment by sodero** [comment]
Great!

**Comment by kas1e** [comment]
Probably for morphos that "unsigned long __stack = 1000000;" should work (i hope, didn't checked)

**Comment by sodero** [comment]
I'll try to make it crash using 4096 and then apply the change and see if it works.

**Comment by kas1e** [comment]
Working from os_amiga.c , i just added it nicely and commit, so once you find out (or check out) way if it works for morphos/aros, you can change it (if necessary, maybe it will works as it already) ).

**Comment by sodero** [comment]
It works! Now I can use the hollywood plugin with a (shell) stack size of 8192 on AROS. Will verify this on MorphOS as well later.

**Comment by kas1e** [comment]
Yeah, seems works then. For sake of interest i think something like snoopdos should show the stack size of currently running process (with all min / max / used sizes) (if i remember right snoopdos works on morphos and aros).

Btw, found at moment new issue which i not remember if we deal with before or not, but at least on previous amigaos4 release version i didn't have that : when i press "iconify" and then "deiconify" back , then whole area of text is not refreshed and staying white till i resize window. Is it the same for you currently with latest codes too ?

**Comment by sodero** [comment]
Hm, yes it behaves in the same way here. Perhaps something with the MUI setup recently, I'll look into it.

**Comment by sodero** [comment]
Ah, now I understand why, this is due to the change of where the off screen bitmap is allocated. It used to be allocated in VimConNew and therefore never freed when VimConCleanup was called (which is happening when iconifying the window), it survived, that's why the contents never disappeared. Now it's allocated and freed in every iconify/show cycle.

**Comment by kas1e** [comment]
So it was some change which was done before migrating to github ? It was one of those attempts to get rid of empty startup window ?

**Comment by sodero** [comment]
Yes, it was done before the migration. It was based on the feedback from Thore. I was violating best practices without knowing it :)

**Comment by kas1e** [comment]
Ah ok, so will wait for commit-fix then :)

**Comment by sodero** [comment]
Fixed. I think.

**Comment by kas1e** [comment]
Well, no. 

First run : i was able to iconify/deiconify it, and it works. Then second time trying iconify : busy loop (or os lockup, dunno).

Second run: even first time can't iconify, everything freezes once i press iconify gadget.

Third run: runing vim, load file, a bit resize window, then trying to iconify : same freeze of everything.

**Comment by kas1e** [comment]
And on some other run, i can iconify, then deiconify : and it the same white as before.

In other words, we need to rollback that commit, as it definately not help, but only make it freezes most of time when you press iconify.

I assume on morphos and aros its the same , right ?

**Comment by sodero** [comment]
:( 

Will dig some more...

**Comment by kas1e** [comment]
Maybe back as it was originally till you change it ? All those iconify and jump to screen issues can be related to whole way how you init/free things, and seens need proper rewrite, but at least as it was it worked

**Comment by sodero** [comment]
The iconification issue is about how control is passed between Vim and the MUI part. The MUI part itself is ok.

**Comment by kas1e** [comment]
Interesting .. I mean , before there wasn't such issue , it appears to be there only when you change the things in mui part. If this "Move bitmap allocation" commit are guilti one which made that new bug, then as i see only mui part was changed there. But i sure not expert here, just wild guesses :)

> The iconification issue is about how control is passed between Vim and the MUI part.

Just to note : the control is passed back fine. I.e. you can press on toolbar after deiconification, you can type letters, just whole rendering of main text area are white. Once you resize manually this window, all start to be fine.

I may try to revert back your "Move bitmap allocation" commit localy and test, if that iconification bug gone or no. If there needs for , of course.

**Comment by sodero** [comment]
Of course you're right about this being something that was introduced with the changes in the MUI part. It's just that now the MUI part follows the advice from Thore, which is good (sort of). If this would be a typical Amiga application, it would behave better now. It's just that it's Vim, and the older less "kosher" solution was a better fit :)

Just like you suggest, I will go back to doing the bitmap things the way they were done before. I have a feeling that that is an easier path than trying to bend Vim to our will.

About the control; it's not about user interaction, it's about passing information from the two (or rather 3 if you include MUI) main loops, Vim has it's own that the event driven MUI part must play nicely with. There is a power struggle going on there :)

**Comment by sodero** [comment]
Now it works the way it used to. Is it ok on OS4 too?

**Comment by kas1e** [comment]
Yes, all fine now. So probabaly we can keep that way for better times when all other important things will be done :)

Now i also add some (Msg)msg casts in 3 places, which cause warnings on my 8.2.0 gcc (at least on os4 setup). Plz, check if it builds for aros/morphos and no ifdefs need it.

In meantime what only left is adding fix for jump-to-new-screen functionality, but that one i need to test a bit more, as previous version seems wasn't fix it fully (i.e. it works first times, then freezes).

**Comment by kas1e** [comment]
Btw, i also made a post on amigans about migrating of project to github and co, maybe (theoreticaly) there can be some other contributors one day :)

At least, anyone can browse whole vim's code now without problems just from the web.

**Comment by sodero** [comment]
Nice! 

I'll try without the ifdefs, perhaps we can also remove the ifdefs around the alpha value in the functions setting foreground and background colors. I'll verify that later. It's nice if we can keep it as general as possible.

Perhaps the screen jump problem is caused by the bitmap problem? I'll try one more time to fix it. I'll push my work to a branch to not break things again. Is it possible to block the screen jumping in the meantime (since it can cause crashes)?

**Comment by kas1e** [comment]
Not sure that screen jumping can be bitmap problem. Its jumps, all seems fine, just not fully works as expected: i.e. for example when i put at end of VimConSetup that:
~~~
my->state |= MUIV_VimCon_GetState_Input; 
~~~
(right before return), then, whole jump to screen funcionality start to work, just in some conditions still fail (maybe just need adding the same string in some other place, i need to check that).

For example to reproduce issue when that string added only to VimConSetup, you can do this:

run vim and choice in mui gadget "jump to Workbench" : it jump on the same screen fine at this moment (as expected). Then, again choice the same "jump to Workbench" : at this point, it should do nothing, but still operate correctly. In our case : it then busy loops everything (but not crashes).

So my bet, is that this string need to be added to other functions you mention before as well .. I will check it now.

**Comment by kas1e** [comment]
Or you can reproduce issue (but be sure you have at end of VimConSetup that line above), but just running vim, choice "jump to new screen" , and then, on new vim screen again choice "jump to vim.1" screen (i.e. the same screen where it placed) , and it will busy loops the same as when you 2 times tries to jump on workbench screen.

I am almost sure its about the same busy loop mean that this string need somewhere else, will check now

**Comment by kas1e** [comment]
Well.. nope, putting the same string to VimConMixMax change nothing in that regard ..

**Comment by sodero** [comment]
I pushed a fix to the fix/jump_to_screen branch. Does that fix work on OS4?

**Comment by kas1e** [comment]
Sadly, but no, nothing changes.

I.e. jump-to-screen functionality start to work without those addons, but when you do 2 times "jump to workbench", on second time it busy looping.  

It also can be not just "jump to workbench" 2 times, its just it easy to reproduce with it (on morphos will be "jump to ambiend" on aros "jump to wandarer" or "jump to dopus" probably) . But you can reproduce it also with just "jump to new screen" , and then, on new screen choice "jump to VIM.1" (i.e. on the same screen).

In other words, issue is that 2 times jumping on the same screen cause busy looping. Just for sake of tests i recked about 5 other mui application : they all reacts fine in that terms, so its not mui/os4/etc.

**Comment by kas1e** [comment]
And just to clarify : its not crash, not os freeze, its just busy looping of vim. All other apps at the same time continue to works fine, i can switch between screens via amiga+m, run other apps, etc. I even can back to the shell window from which i run vim , and press "ctrl+c", so it will closed vim fine and exit from.

So something somewhere doing nasty things .. Maybe its some "break" forgotten somewhere ? When i compile gui_mui.c , i have warning of such kind:

~~~
gui_mui.c: In function ‘VimConHandleEvent’:
gui_mui.c:1747:12: warning: this statement may fall through [-Wimplicit-fallthrough=]
         if (VimConMouseHandleButton (cls, obj, msg))
            ^
gui_mui.c:1752:5: note: here
     default:
     ^~~~~~~
~~~

**Comment by sodero** [comment]
Oh, I didn't try that. I just jumped back and forth many times. Will try that.

**Comment by kas1e** [comment]
Yeah back and forth in that way works and as before we fix it with one line, and in your "screen-to-jump" branch fix. Probabaly that fix you do good as well (as more clean), just need to understand wtf happens with that double-repeating. Its like it loose some flag after second try or something. Or maybe skip setting of some flag

**Comment by sodero** [comment]
Yes, it's weird. I can reproduce it now at least. I'll do some more digging, but unfortunately someone wants me to leave the computer now :)

**Comment by sodero** [comment]
Found something strange. When jumping to the same screen MUI wants us to wait for 0. I added a protection for that (but I have no clue why this happens). With this fix / workaround, jumping to the same screen works on MOS. Does it work on OS4?

**Comment by sodero** [comment]
BTW, I need to do some radical work on this repo. The way I set this up is a really bad idea, all the clean ups in the beginning make it hard to track upstream Vim. I need to go back and re-apply our patches. The end result will be the same except that there will be a lot of files that aren't relevant to us, but it will make it a lot easier later on.

When I'm done with this, we need to do a fresh clone off the repository. Let me know when this is OK for you. If you have things that you haven't commited yet, I'll wait for you, no hurry :)

**Comment by sodero** [comment]
And of course, if I mess up, there's a backup here:

https://sourceforge.net/projects/mui-vim/

**Comment by kas1e** [comment]
Yes, everything works fine now  ! I can ask Thore why mui wait for 0 there (just so we will know details).

The only thing that is not updated at moment from all the previous tests and fixes we done , its getting rid of home: assignment, by changing in  os_amiga.h that:

~~~
#ifndef DFLT_RUNTIMEPATH
# define DFLT_RUNTIMEPATH "home:vimfiles,$VIM/vimfiles,$VIMRUNTIME,$VIM/vimfiles/after,home:vimfiles/after"
#endif
~~~

On that (just copy of CLEAN_RUNTIMEPATH):

~~~
#ifndef DFLT_RUNTIMEPATH
# define DFLT_RUNTIMEPATH "$VIM/vimfiles,$VIMRUNTIME,$VIM/vimfiles/after"
#endif
~~~

But that one can wait easy till you rebase repo. I.e. do what you think you need to do now, and once you done, we can made that change  and continue :)

**Comment by sodero** [comment]
Very oood! I'll try to do the repo work tonight.

**Comment by sodero** [comment]
The repo has been rebuilt. All of the files from upstream are included now. Less clean, but a lot easier to work with upstream Vim. I did a small minor version bump just to try it out.

You should be able to build from src just like before. I tried AROS and MOS and they work. Let me know if I missed something OS4 specific, but I don't think I have, I have been very thorough ;)

You need to do a fresh clone otherwise things will go €#"/)"#%.

**Comment by sodero** [comment]
BTW, I also added a placeholder in CREDITS.txt and gui_mui.c where you can put your name, if you want to (otherwise I can remove it).

**Comment by kas1e** [comment]
Yeah, everything builds fine with reorganized repo, cool ! Tested also resulting binary: all seems the same fine as before. Iconify works, jump to screen works, etc ,etc.

I also got the answer from Thore, about that mistery waiting for 0 (about that dbl-jump on the same screen), and that what he answer:


> You are initializing your "sig" variable at the wrong point. The typical MUI event loop looks like this:


```C
ULONG sigs = 0;

while ((LONG)DoMethod(app,MUIM_Application_NewInput,&sigs) != MUIV_Application_ReturnID_Quit)
{
     if (sigs != 0)
     {
         sigs = Wait(sigs | SIGBREAKF_CTRL_C);
         if (sigs & SIGBREAKF_CTRL_C)
         {
             break;
         }
     }
}
```

> Note that "sigs" is initialized *once* outside the event loop and each value "sigs" that is returned 
from the Wait() call is passed in again to the next call of the MUIM_Application_NewInput method. 
This is required to let MUI know which of its own internal signals have occured and to let MUI react 
on its own internal message ports.

> However, your loop looks like this:

```C
while(do_not_bail_out)
{
    ULONG sigs = 0;

    if ((LONG)DoMethod(app,MUIM_Application_NewInput,&sigs) != MUIV_Application_ReturnID_Quit)
    {
        if (sigs != 0)
        {
            sigs = Wait(sigs | SIGBREAKF_CTRL_C);
            if (sigs & SIGBREAKF_CTRL_C)
            {
                // bail out
                break;
            }
        }
    }
}
```

> The MUIM_Application_NewInput method always gets a zero "sigs" value passed in and hence MUI cannot  react on any of its own internal signals as there none.

> Since the "do_not_bail_out" loop of VIM in gui_mch_wait_for_chars () is called whenever it has 
nothing else to do the value of "sigs" returned by MUIM_Application_NewInput must be stored 
somewhere else than on the stack to be able to pass it in again on the next call. Since there is 
nothing like a context structure a function static variable might be sufficient already. That way 
the variable is initialized only once during program launch, is visible inside the function only and 
keeps it value even across several function calls. Not a really nice approach, but sometimes it 
cannnot be avoided.

Hope it will be of any help :)

Also last thing need to commit is that "home:" assignment change (will do that today)

As for credits, that sure not that important as from my side there wasn't any code changes, only some small fixes and os4 ifdefs :) But yeah, will change it later too.

**Comment by sodero** [comment]
Excellent, Thore is a great resource. I'm not to familiar with MUI to be honest, this was my first MUI application :)

**Comment by kas1e** [comment]
Hi !

Tested 8 hours ago pre-release : all seems works on os4 as it. Through, there is left that "home:" assignment which was the last thing we didn't put in the repo from the old-first tests. I.e. there was some change in some file where we remove some tihng to make that home: asknowlede disappear. 

Can you remind plz where it hosted before (where we have that long disussion about os4 port, before you swith to github) , so i can find what change we do.

But in terms of icons, whole archive packing : all seems correct. It also builded  with clib2 as need it , so seems everytihng as should be in other terms.

Through there still that damn empty-output window on start, but that was like this even before, so not a problem then.

**Comment by sodero** [comment]
Hi kas1e!

Nice to hear that it's working. I initially tried to do some OS4 work using Amiga Forever and OS4 classic, but it was painfully slow (that darn Tabor needs to be released ASAP), so I gave up and  blindly set up a cross compiler in Travis that builds and deploys to GitHub automatically, a bit of a hack but it seems to work, pretty neat in fact:

https://travis-ci.org/sodero/MUI-Vim

I really like Marlons effort to create a Docker image for cross compilers for *all* Amiga flavours, great initiative, we need more of that:

https://github.com/AmigaPorts/docker-amiga-gcc

The discussion about HOME was here:

https://sourceforge.net/p/spaden/tickets/1/

I don't know why this change was lost but feel free to add it again.

I'm afraid that I reintroduced the installer (I know that everyone at amigans.net hate it ;), but I got so depressed by seeing the Amiga Future review where VIM: obviously was wrong, the toolbar graphics was missing and probably VIMRC too.

**Comment by kas1e** [comment]
That change just wasn't added when i upload changes there, was in hope to add it together with get-rid of empty window ..

As for intstaller , i think that better not to make installer back, but instead, make autoassignment of VIM: right in the VIM's code. I.e. so there will be no needs for any assigns ever. I can help with that, so we can see how to make it for morphos, aos and aros. By this way there for first will be no needs for any assignemnts (good), and there will be no needs for any installers (which is good too).

As for AF , don't count much on them, they test software in pretty bad way some times. But not that bad as Amitopia of course, but still :)

**Comment by sodero** [comment]
Sure, an auto assign would be good, but I think that we should have a check to see if it's already assigned in case anyone would like to separate the binary from the runtime, which was the upstream idea from the beginning. In general I think we should strive for not making to intrusive changes. I try to integrate as much as possible in the mainline to make sure that they are aware of that Vim is still used on Amiga. That will reduce the risk of the Amiga parts being removed completely. There have been discussions about this, but so far Bram has been a good friend of ours :)

**Comment by kas1e** [comment]
Aha got it .. 

Can you also add to repo that change about home: assignment we have on os4 (but that as i remember didn't hurt any other os).

I.e. we just change that :

```
#ifndef DFLT_RUNTIMEPATH
# define DFLT_RUNTIMEPATH "home:vimfiles,$VIM/vimfiles,$VIMRUNTIME,$VIM/vimfiles/after,home:vimfiles/after"
#endif
```
on that:

```
#ifndef DFLT_RUNTIMEPATH
# define DFLT_RUNTIMEPATH "$VIM/vimfiles,$VIMRUNTIME,$VIM/vimfiles/after"
#endif
```

(its just copy of CLEAN_RUNTIMEPATH).

So with that one we will have all the same as it was with os4 build on os4depot.

**Comment by sodero** [comment]
Yep, fixed it. Building right now https://www.travis-ci.org/sodero/MUI-Vim/builds/599842672
Artefacts will be up in 5 minutes or so.

**Comment by kas1e** [comment]
Tested builds which were uploaded on os4depot : all fine for os4. That includes the installer and all the things like icons and stuff. Uploaded on os4depot as well a new version and that ticket we can close probably.

Thanks for maintaining it all!

**Comment by sodero** [comment]
My pleasure, thanks for the initial port!

**Comment by kas1e** [comment]
Btw, weren't you able to setup aos4 somewhere ?:) Just found on os4depot one comment from user, who says "is it possible to get rid of iconx window started when we run gvim" , i.e. that one which we had before and which is seems didn't spawns on morphos/aros (?). Want to try to get rid of it too.

**Comment by sodero** [comment]
I've got Amiga Forever + OS4 classic but it's painfully slow and cross-compiling on that setup is tedious.

I bought AAA the other day though, so hopefully I'll have a Tabor to play with within the near future (depending on how you define near).
